src/librustc/ty/context.rs

```diff
@@ -2867,14 +2867,18 @@ impl<T, R, E> InternIteratorElement<T, R> for Result<T, E> {
         // lower bounds from `size_hint` agree they are correct.
         Ok(match iter.size_hint() {
             (1, Some(1)) => {
-                f(&[iter.next().unwrap()?])
+                let t0 = iter.next().unwrap()?;
+                assert!(iter.next().is_none());
+                f(&[t0])
             }
             (2, Some(2)) => {
                 let t0 = iter.next().unwrap()?;
                 let t1 = iter.next().unwrap()?;
+                assert!(iter.next().is_none());
                 f(&[t0, t1])
             }
             (0, Some(0)) => {
+                assert!(iter.next().is_none());
```

Comment:
If it matters this could be written without unwrap's or asserts.

```rust
let a = match iter.next() {
    None => return Ok(f(&[])),
    Some(a) => a?,
};
let b = match iter.next() {
    None => return Ok(f(&[a])),
    Some(b) => b?,
};
let c = match iter.next() {
    None => return Ok(f(&[a, b])),
    Some(c) => c?,
};
let mut vec: SmallVec<[_; 8]> = SmallVec::with_capacity(iter.size_hint().0 + 3);
vec.push(a);
vec.push(b);
vec.push(c);
for result in iter {
    vec.push(result?);
}
Ok(f(&vec))
```

Reply:
@RalfJung Should we perf test this variant?

Reply:
No opinion. @nnethercote what do you think?

Reply:
The `vec` construction could be more concise, e.g. using `extend`. And I would use `t0`, `t1`, `t2` instead of `a`, `b`, `c`. And the comment needs updating. Otherwise, fine by me to use this form if the perf isn't hurt.